node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -283,37 +306,61 @@ impl Initialized {
 		self.participation.process_active_leaves_update(ctx, &update).await?;
 
 		if let Some(new_leaf) = update.activated {
-			match self
-				.rolling_session_window
-				.cache_session_info_for_head(ctx.sender(), new_leaf.hash)
-				.await
-			{
-				Err(e) => {
-					gum::warn!(
+			let session_idx =
+				self.runtime_info.get_session_index_for_child(ctx.sender(), new_leaf.hash).await;
+
+			match session_idx {
+				Ok(session_idx)
+					if self.last_consecutive_cached_session.is_none() ||
+						session_idx >
+							self.last_consecutive_cached_session.expect(
+								"The first clause explicitly handles `None` case. qed.",
+							) =>
+				{
+					// There is a new session. Perform a dummy fetch to cache it.
+					let mut gap_in_cache = false;
+					for idx in self
+						.last_consecutive_cached_session
+						.unwrap_or(session_idx.saturating_sub(DISPUTE_WINDOW.get() - 2)) +
```

Comment:
I think you subtracted 1 from `DISPUTE_WINDOW.get() - 1` instead of adding it (because of double negation)
```suggestion
						.unwrap_or(session_idx.saturating_sub(DISPUTE_WINDOW.get())) +
```

Reply:
These substractions ... -1, -2  make me a bit uneasy :)

Reply:
> I think you subtracted 1 from DISPUTE_WINDOW.get() - 1 instead of adding it (because of double negation)

Yes, it should be `session_idx - (DISPUTE_WINDOW - 1) - 1`. `saturating_sub` messed up the arithmetic in my brain.

Maybe replacing `last_consecutive_cached_session` with a `bool` will simplify the logic as you suggested. We might do a few extra calls but they will be cached, so won't matter that much.